File: vertx-elasticsearch/src/main/java/fr/myprysm/vertx/elasticsearch/action/get/GetRequest.java
package fr.myprysm.vertx.elasticsearch.action.get;

import fr.myprysm.vertx.elasticsearch.action.BaseRequest;
import fr.myprysm.vertx.elasticsearch.action.support.FetchSourceContext;
import io.vertx.codegen.annotations.DataObject;
import io.vertx.core.json.JsonObject;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.experimental.Accessors;
import org.elasticsearch.common.lucene.uid.Versions;
import org.elasticsearch.index.VersionType;

import java.util.List;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Accessors(chain = true)
@EqualsAndHashCode(callSuper = true)
@DataObject(generateConverter = true)
public class GetRequest extends BaseRequest {
    public static final String DEFAULT_TYPE = "_all";

    private String index;
    private String type = DEFAULT_TYPE;
    private String id;
    private String routing;
    private String parent;
    private String preference;

    private List<String> storedFields;

    private FetchSourceContext fetchSourceContext;

    private boolean refresh = false;
    private boolean realTime = true;
    private VersionType versionType = VersionType.INTERNAL;
    private long version = Versions.MATCH_ANY;

    public GetRequest(JsonObject json) {
        super(json);
        GetRequestConverter.fromJson(json, this);
    }

    public GetRequest(String index, String type, String id) {
        this.index = index;
        this.type = type;
        this.id = id;
    }

    @Override
    public GetRequest addHeader(String headerKey, String headerValue) {
        return (GetRequest) super.addHeader(headerKey, headerValue);
    }

    public JsonObject toJson() {
        JsonObject json = super.toJson();
        GetRequestConverter.toJson(this, json);
        return json;
    }
}

File: vertx-elasticsearch/src/main/kotlin/fr/myprysm/vertx/elasticsearch/kotlin/action/get/GetRequest.kt
package fr.myprysm.vertx.elasticsearch.kotlin.action.get

import fr.myprysm.vertx.elasticsearch.action.get.GetRequest
import fr.myprysm.vertx.elasticsearch.action.support.FetchSourceContext
import org.elasticsearch.index.VersionType

fun GetRequest(
  fetchSourceContext: fr.myprysm.vertx.elasticsearch.action.support.FetchSourceContext? = null,
  headers: Map<String, String>? = null,
  id: String? = null,
  index: String? = null,
  parent: String? = null,
  preference: String? = null,
  realTime: Boolean? = null,
  refresh: Boolean? = null,
  routing: String? = null,
  storedFields: Iterable<String>? = null,
  type: String? = null,
  version: Long? = null,
  versionType: VersionType? = null): GetRequest = fr.myprysm.vertx.elasticsearch.action.get.GetRequest().apply {

  if (fetchSourceContext != null) {
    this.setFetchSourceContext(fetchSourceContext)
  }
  if (headers != null) {
    this.setHeaders(headers)
  }
  if (id != null) {
    this.setId(id)
  }
  if (index != null) {
    this.setIndex(index)
  }
  if (parent != null) {
    this.setParent(parent)
  }
  if (preference != null) {
    this.setPreference(preference)
  }
  if (realTime != null) {
    this.setRealTime(realTime)
  }
  if (refresh != null) {
    this.setRefresh(refresh)
  }
  if (routing != null) {
    this.setRouting(routing)
  }
  if (storedFields != null) {
    this.setStoredFields(storedFields.toList())
  }
  if (type != null) {
    this.setType(type)
  }
  if (version != null) {
    this.setVersion(version)
  }
  if (versionType != null) {
    this.setVersionType(versionType)
  }
}

